awkward-cpp/include/awkward/io/json.h

```diff
@@ -135,20 +135,91 @@ namespace awkward {
     inline int64_t find_key(const char* str) noexcept {
       int64_t* offsets = string_offsets_.data();
       char* chars = characters_.data();
+      int64_t i;
+      int64_t j;
       int64_t stringi;
       int64_t start;
       int64_t stop;
-      for (int64_t i = current_instruction_ + 1;  i <= current_instruction_ + argument1();  i++) {
+      uint64_t chunkmask;
+      // optimistic: fields in data are in the order specified by the schema
+      if (argument1() != 0) {
+        // increment the current (last seen) field with wrap-around
+        record_current_field_[argument2()]++;
+        if (record_current_field_[argument2()] == argument1()) {
+          record_current_field_[argument2()] = 0;
+        }
+        j = record_current_field_[argument2()];
+        // use the record_current_field_ (as j)
+        i = current_instruction_ + 1 + j;
         stringi = instructions_.data()[i * 4 + 1];
         start = offsets[stringi];
         stop = offsets[stringi + 1];
         if (strncmp(str, &chars[start], (size_t)(stop - start)) == 0) {
-          return instructions_.data()[i * 4 + 2];
+          // ensure that the checklist bit is 1
+          chunkmask = (uint64_t)1 << (j & 0x3f);
+          if ((record_checklist_[argument2()][j >> 6] & chunkmask) == 0) {
+            return -1;  // ignore the value of a duplicate key
+          }
+          // set the checklist bit to 0
+          record_checklist_[argument2()][j >> 6] &= ~chunkmask;
+          return key_instruction_at(i);
+        }
+      }
+      // pessimistic: try all field names, starting from the first
+      for (i = current_instruction_ + 1;  i <= current_instruction_ + argument1();  i++) {
+        // not including the one optimistic trial
+        if (i != current_instruction_ + 1 + record_current_field_[argument2()]) {
+          stringi = instructions_.data()[i * 4 + 1];
+          start = offsets[stringi];
+          stop = offsets[stringi + 1];
+          if (strncmp(str, &chars[start], (size_t)(stop - start)) == 0) {
+            // set the record_current_field_
+            j = i - (current_instruction_ + 1);
+            record_current_field_[argument2()] = j;
+            // ensure that the checklist bit is 1
+            chunkmask = (uint64_t)1 << (j & 0x3f);
+            if ((record_checklist_[argument2()][j >> 6] & chunkmask) == 0) {
```

Comment:
Could this be `key_already_filled(argument2(), j)`?

Reply:
It wouldn't get to reuse the `chunkmask`. (Not a big deal, but this is a hot loop in an inline function.)